common/transport/amqp/devdoc/sender_link_requirements.md

```diff
@@ -0,0 +1,78 @@
+
+### constructor(linkAddress: string, linkOptions: any, amqp10Client: amqp10.AmqpClient)
+
+**SRS_NODE_AMQP_SENDER_LINK_16_002: [** The `SenderLink` class shall inherit from `EventEmitter`. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_003: [** The `SenderLink` class shall implement the `AmqpLink` interface. **]**
+
+### attach(callback: (err?: Error) => void): void
+
+**SRS_NODE_AMQP_SENDER_LINK_16_004: [** The `attach` method shall use the stored instance of the `amqp10.AmqpClient` object to attach a new link object with the `linkAddress` and `linkOptions` provided when creating the `SenderLink` instance. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_022: [** The `attach` method shall call the `callback` if the link was successfully attached. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_006: [** The `SenderLink` object should subscribe to the `detached` event of the newly created `amqp10` link object. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_007: [** The `SenderLink` object should subscribe to the `errorReceived` event of the newly created `amqp10` link object. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_008: [** If the `amqp10.AmqpClient` fails to create the link the `callback` function shall be called with this error object. **]**
+
+### detach(): void
+
+**SRS_NODE_AMQP_SENDER_LINK_16_009: [** The `detach` method shall detach the link created by the `amqp10.AmqpClient` underlying object. **]**
+
+### send(message: AmqpMessage, callback: (err?: Error, result?: results.MessageEnqueued) => void): void
+
+**SRS_NODE_AMQP_SENDER_LINK_16_010: [** The `send` method shall use the link created by the underlying `amqp10.AmqpClient` to send the specified `message` to the IoT hub. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_011: [** If the state machine is not in the `attached` state, the `SenderLink` object shall attach the link first and then send the message. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_012: [** If the message cannot be sent the `callback` shall be called with an `Error` object describing the AMQP error reported by the service. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_013: [** If the message is successfully sent, the `callback` shall be called with a first parameter (error) set to `null` and a second parameter of type `MessageEnqueued`. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_014: [** If the link is detached while a message is being sent, the `callback` shall be called with an `Error` object describing the AMQP error that caused the detach to happen in the first place. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_019: [** While the link isn't attached, the messages passed to the `send` method shall be queued. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_020: [** When the link gets attached, the messages shall be sent in the order they were queued. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_021: [** If the link fails to attach and there are messages in the queue, the callback for each message shall be called with the error that caused the detach in the first place. **]**
```

Comment:
>callback for each message shall be called [](start = 109, length = 41)

I assume the messages get removed from the queue in this case, right?  Should that be spelled out here?

Reply:
the fact that there is a queue in internal implementation details imho, but I'm not against spelling it out if you think it's better

---
In reply to: [139495305](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139495305) [](ancestors = 139495305)

Reply:
Maybe this is intuitive, but I'm talking about the fact that caller will have to re-submit the message if the link fails to attach.  The AMQP object doesn't save the message to send it the next time the link attaches.

---
In reply to: [139537136](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139537136) [](ancestors = 139537136,139495305)

Reply:
Ohh I see. well yes the idea with these changes is that the retry logic itself will be handled at the client level (which in that case is the caller).

---
In reply to: [139806076](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139806076) [](ancestors = 139806076,139537136,139495305)